src/libpmemset/sds.c

```diff
@@ -0,0 +1,351 @@
+ */
+int
+pmemset_sds_unregister_record(struct pmemset_sds_record *sds_record)
+{
+	int ret = 0;
+	struct pmemset_sds *sds = sds_record->sds;
+
+	util_rwlock_wrlock(&State.sds_lock);
+	struct ravl_node *node = ravl_find(State.rtree, sds_record,
+			RAVL_PREDICATE_EQUAL);
+	if (!node) {
+		ret = -ENOENT;
+		goto err_unlock;
+	}
+
+	ravl_remove(State.rtree, node);
+
+	sds->refcount -= 1;
+	ASSERT(sds->refcount >= 0);
+
+err_unlock:
+	util_rwlock_unlock(&State.sds_lock);
+	return ret;
+}
+
+/*
+ * pmemset_sds_new -- allocates and initialize SDS structure
+ */
+int
+pmemset_sds_new(struct pmemset_sds **sds_ptr, struct pmemset_source *src)
+{
+	LOG(3, "sds_ptr %p src %p", sds_ptr, src);
+
+	*sds_ptr = NULL;
+	struct pmemset_file *part_file = pmemset_source_get_set_file(src);
+	struct pmem2_source *pmem2_src =
+			pmemset_file_get_pmem2_source(part_file);
+
+	int ret = 0;
+	struct pmemset_sds *sds = pmemset_malloc(sizeof(*sds), &ret);
+	if (ret)
+		return ret;
+
+	memset(sds->id, 0, PMEMSET_SDS_DEVICE_ID_LEN);
+
+	/* read device unsafe shutdown count */
+	ret = pmem2_source_device_usc(pmem2_src, &sds->usc);
+	if (ret) {
+		if (ret == PMEM2_E_NOSUPP)
+			ret = PMEMSET_E_SDS_ENOSUPP;
+		goto err_free_sds;
+	}
+
+	/* read device ID length */
+	size_t len = 0;
+	ret = pmem2_source_device_id(pmem2_src, NULL, &len);
+	if (ret) {
+		if (ret == PMEM2_E_NOSUPP)
+			ret =  PMEMSET_E_SDS_ENOSUPP;
+		goto err_free_sds;
+	}
+
+	ASSERT(len <= PMEMSET_SDS_DEVICE_ID_LEN);
+
+	/* read device ID */
+	ret = pmem2_source_device_id(pmem2_src, (char *)&sds->id, &len);
+	if (ret) {
+		if (ret == PMEM2_E_NOSUPP)
+			ret = PMEMSET_E_SDS_ENOSUPP;
+		goto err_free_sds;
+	}
+
+	sds->refcount = 0;
+
+	*sds_ptr = sds;
+
+	return 0;
+
+err_free_sds:
+	Free(sds);
+	return ret;
+}
+
+/*
+ * pmemset_sds_delete -- deallocate sds structure
+ */
+int
+pmemset_sds_delete(struct pmemset_sds **sds_ptr)
+{
+	Free(*sds_ptr);
+	*sds_ptr = NULL;
+	return 0;
+}
+
+/*
+ * pmemset_sds_is_initialiized -- checks if the SDS was initialized
+ */
+static bool
+pmemset_sds_is_initialiized(const struct pmemset_sds *sds_old,
+	const struct pmemset_sds *sds_new)
+{
+	return strncmp(sds_old->id, sds_new->id,
+			PMEMSET_SDS_DEVICE_ID_LEN) == 0;
+}
+
+/*
+ * pmemset_sds_is_consistent -- checks if the SDS indicates possible data
+ *                              corruption
+ */
+static bool
+pmemset_sds_is_consistent(const struct pmemset_sds *sds_old,
+		const struct pmemset_sds *sds_new)
+{
+	return sds_old->usc == sds_new->usc;
+}
+
+/*
+ * pmemset_sds_state_check_and_possible_refresh -- checks part state and refresh
+ *                                                 usc if it's just outdated
+ */
+int
+pmemset_sds_state_check_and_possible_refresh(struct pmemset_source *src,
+		enum pmemset_part_state *state_ptr)
+{
+	struct pmemset_sds *sds = pmemset_source_get_extras(src).sds;
+
+	struct pmemset_sds *sds_new;
+	int ret = pmemset_sds_new(&sds_new, src);
```

Comment:
so you are allocating a few hundred-byte struct on the heap to just free it at the end of the function?
I know this is not performance-sensitive code, but come on ;)